### DISCUSSION TITLE
problem is testCantLiquidateGoodHealthFactor function

### DISCUSSION THREAD

**Discussion by khushbusandhu** [Cyfrin/foundry-full-course-cu#703] [General]
here's the function 
```
 function testCantLiquidateGoodHealthFactor() public DepositeCollateralAndMintedDsc {
        ERC20Mock(weth).mint(LIQUIDATOR, COLLATERAL_TO_COVER);
        vm.startPrank(LIQUIDATOR);
        ERC20Mock(weth).approve(address(dsce), COLLATERAL_TO_COVER);
        dsce.depositCollateralAndMintDsc(weth, COLLATERAL_TO_COVER, AMOUNT_TO_MINT);
        dsc.approve(address(dsce), AMOUNT_TO_MINT);
        vm.expectRevert(DSCEngine.DSCEngine__HealthFactorOk.selector);
        dsce.liquidate(weth, USER, AMOUNT_TO_MINT);
        vm.stopPrank();
    }
```
Why we approved from dsc even if we have improved token from ERC20?

**Comment by theirrationalone** [comment]
ERC20Mock(weth).approve(address(dsce), COLLATERAL_TO_COVER):::::::>

ERC20Mock:
Heyy DSCEngine, i promise to pay you x amount of collateral. Please account that amount for my owner(user) 

Deposited and minted, Promise fulfilled.... 

dsc.approve(address(dsce), AMOUNT_TO_MINT);

Dsc:
Heyy DSCEngine, i promise to pay you back x amount of dsc, that my owner borrowed from you and please give related amount of collateral back to my owner.

**Comment by khushbusandhu** [comment]
okay I got it
thanks @theirrationalone for explaining in such a interesting way

**Comment by YapBarry** [comment]
My understanding is that "ERC20Mock(weth).approve" is necessary because the Liquidator is depositing weth into DSCEngine in order to mint DSC. "ERC20Mock(weth).approve" allows DSCEngine to transfer the weth from the Liquidator to DSCEngine.

After depositing weth and minting DSC, the Liquidator is now going to try to call the liquidate function to liquidate the "user" who might be having a health factor below minimum health factor and receiving the weth collateral that the "user" deposited.

dsc.approve is necessary because if the liquidator wants to liquidate the "user", the liquidator needs to send his DSC to the DSCEngine (to help repay the user's DSC debt) and then DSCEngine will transfer the weth collateral deposited by the "user" to the liquidator. 

Hope this makes sense! Though I am probably 1 year late xD

**Reply by EngrPips** [reply]
Interesting, Thanks for the explanation.
